### PR TITLE
Flag for WebGLTile Layers for blending with gl.ONE

### PR DESCRIPTION
**Pull request by Punzo**
**Context**:

A digital slide microscopy images viewer based on OpenLayer ([link](https://github.com/MGHComputationalPathology/dicom-microscopy-viewer) used by this App [Slim](https://github.com/MGHComputationalPathology/slim)). The images are pyramidal multi resolution layers. Data shared in the context of Image Data Commons of the National Cancer Institute ([IDC](https://imaging.datacommons.cancer.gov/)).

**Why**: 

we need to blend with as gl.blendFunc(gl.ONE, gl.ONE) only for a subset of Tile Layers (WebGL).

**What**:

have a flag in the TileLayer (WebGL) style to change the webgl blendfunction. 

**Long explanation**:

we use OpenLayer to visualize microscopy images.  In our use case, we need to color and blend images from different TileLayers (https://www.youtube.com/watch?v=9VfrP6CM3nA).

We were performing the rendering with an webgl offscreen render.  With the new introduction of WebGL Tile Layers, we are switching from the offscreen render to a fully OpenLayer implementation (we have already tested that we get a major boost in performances, which is great!).

One thing that we miss in the new API, it is a parameter to set different blending options for the webgl rendering in OpenLayer (i.e. in https://github.com/openlayers/openlayers/blob/main/src/ol/webgl/Helper.js#L485-L488)

------------------------------------------------------------------------------------------

This PR is a draft and just to point out what we would like to have, and we would like to agree (and then contribute) on some  general API interface.

### CODE REVIEW

**Comment by hackermd** [comment]
For our use case, we only need additive blending, but it may be interesting to support additional blending modes in OpenLayers.

For the Canvas renderer, the composition can be configured via the [CanvasRenderingContext2D.globalCompositeOperation](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/globalCompositeOperation). Would it make sense to allow users to configure blending behaviour (for both the Canvas and the WebGL renderer) using a `compositeOperation` parameter (see [Composition examples](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Compositing/Example)). The argument could then be mapped internally to the corresponding parameters of `glBlendEquation()` and `glBlendFunc()`.

```javascript
prepareDraw(frameState, opt_disableAlphaBlend, compositeOperation) {
  ...
  if (compositionOperation === "lighter") {
    glBlendEquation(GL_FUNC_ADD)
  }
  ...
)
```

**Comment by Punzo** [comment]
> For our use case, we only need additive blending, but it may be interesting to support additional blending modes in OpenLayers.
> 
> For the Canvas renderer, the composition can be configured via the [CanvasRenderingContext2D.globalCompositeOperation](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/globalCompositeOperation). Would it make sense to allow users to configure blending behaviour (for both the Canvas and the WebGL renderer) using a `compositeOperation` parameter (see [Composition examples](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Compositing/Example)). The argument could then be mapped internally to the corresponding parameters of `glBlendEquation()` and `glBlendFunc()`.
> 
> ```js
> prepareDraw(frameState, opt_disableAlphaBlend, compositeOperation) {
>   ...
>   if (compositionOperation === "lighter") {
>     glBlendEquation(GL_FUNC_ADD)
>   }
>   ...
> )
> ```

Hi Markus, 

yes I was thinking something like that, but we would need to set acutally the _blendFunc_. I was thinking to have enum 1:1 with the WebGL variables.

 ```js
 prepareDraw(frameState, opt_disableAlphaBlend, glBlendEquation, glBlendFunc1, glBlendFunc2) {
   ...
    blendEquation(glBlendEquation)
    blendFunc(glBlendFunc1, glBlendFunc2)
   ...
 )
 ```

where 
glBlendEquation: GL_FUNC_ADD, GL_FUNC_SUBTRACT, GL_FUNC_REVERSE_SUBTRACT, GL_MIN,  GL_MAX (in our case we need GL_FUNC_ADD, but it is the default parameter)
glBlendFunc1/glBlendFunc2: https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/blendFunc#constants

so in this way we would have maximum flexibility. I don't like the idea to have hardcoded presets as https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/globalCompositeOperation with a lot of if conditions in prepareDraw.

Anyway, let's see what OpenLayer developer agree with us and what they would like to do.

**Comment by mike-000** [comment]
For 2d canvas renderers a globalCompositeOperation can be easily configured/reset at the layer prerender and postrender events:

`evt.context.globalCompositeOperation = 'lighter';`

I don't think it is necessary to provide any additional functionality for 2d canvas layers or replicate those for WebGL.

**Comment by Punzo** [comment]
> For 2d canvas renderers a globalCompositeOperation can be easily configured/reset at the layer prerender and postrender events:
> 
> `evt.context.globalCompositeOperation = 'lighter';`
> 
> I don't think it is necessary to provide any additional functionality for 2d canvas layers or replicate those for WebGL.

Thanks for you feedback. 

Probably I was not clear on our requirements, let me explain in more details. 

We have greyscale images and the user can apply a color function (rgb color, opacity, min/max, etc..., see screenshot below) to the image interactively. The images are pyramidal and the tiles are also rather large (> 10^3x10^3 pixels). 

Therefore we need fast webgl processing, and we can't use cpu processing with the canvas 2D Tile Layer for coloring the image (it is just too slow, while the blending by setting evt.context.globalCompositeOperation was fast enough) .  

This is the reason why we were using an offscreen webgl rendering for coloring the images, then we were passing the image to the canvas 2D tile layer and finally making the blening in openLayer by setting evt.context.globalCompositeOperation = 'lighter' as you described. This workflows worked well, but there was a bottleneck in passing the rendered image as a png from the offscreen render to the canvas 2D tile layer, and that's why we are migrating to the new WebGL Tile Layer with OpenLayer version 6.9.

Blending capabilities for the WebGL Tile Layer for me looks a general use cases and many users could benefit from such API. If you still think that it is not worth having such API, we will fork OpenLayer. However, we really would like to avoid forking for a small change like this one ans we hope to agree on some common blending API for teh WebGL Tile Layer to integrate in OpenLayer.

**Comment by mike-000** [comment]
I have no problem with the general use proposal using WebGL variables.  I don't think any new options are needed for 2d rendering, and therefore hardcoding `globalCompositeOperation` values to match seems inappropriate in a WebGL setting.

Browsers do use some hardware acceleration for 2d canvas operations, but in my experience it is buggy - there is currently an issue with https://openlayers.org/en/latest/examples/disable-image-smoothing.html in Chrome and Edge (only with hardware acceleration enabled) where you may notice some pixels in the smoothing disabled map show incorrect negative elevations.

**Comment by Punzo** [comment]
> I have no problem with the general use proposal using WebGL variables. I don't think any new options are needed for 2d rendering, and therefore hardcoding `globalCompositeOperation` values to match seems inappropriate in a WebGL setting.
> 

I agree with you in not using hardcoded presets as the one from globalCompositeOperation. I will update the PR to pass directly the WebGL variables as I suggested in https://github.com/openlayers/openlayers/pull/12985#issuecomment-965709612. The idea is that the extenal app set up the WebGL blending variables in the WebGL Tile Layer style for each tile layer.

> Browsers do use some hardware acceleration for 2d canvas operations, but in my experience it is buggy - there is currently an issue with https://openlayers.org/en/latest/examples/disable-image-smoothing.html in Chrome and Edge (only with hardware acceleration enabled) where you may notice some pixels in the smoothing disabled map show incorrect negative elevations.

I agree, for the moment image processing with 2d canvas is still not optimal and they do not offer enough flexibility as a pure WebGL approach for specific task.

**Comment by jahow** [comment]
Wouldn't it be possible to use the webgl context available in `prerender` and `postrender` events to do that? See https://github.com/openlayers/openlayers/pull/12933

**Comment by Punzo** [comment]
> Wouldn't it be possible to use the webgl context available in `prerender` and `postrender` events to do that? See #12933

I didn't see that I could have access to the webgl context in the prerender and postrender events. I will try to change the blending vars there. Thanks for pointing out!

**Comment by hackermd** [comment]
@mike-000 @jahow I may have stumbled upon a more fundamental problem. It appears every layer gets rendered into a separate context.

Every layer creates a separate `WebGLTileLayerRenderer`:
https://github.com/openlayers/openlayers/blob/0e19c9aa2b05eb9b28b502aa8d3cd590dabacafa/src/ol/layer/WebGLTile.js#L303-L317

Every renderer creates a separate `WebGLHelper`:
https://github.com/openlayers/openlayers/blob/080fe8ca67da57202ce024ea0267f1e5c8e8a9fa/src/ol/renderer/webgl/Layer.js#L77-L80

Every helper creates a separate canvas/context:
https://github.com/openlayers/openlayers/blob/080fe8ca67da57202ce024ea0267f1e5c8e8a9fa/src/ol/webgl/Helper.js#L255
https://github.com/openlayers/openlayers/blob/080fe8ca67da57202ce024ea0267f1e5c8e8a9fa/src/ol/webgl/Helper.js#L263

**Comment by mike-000** [comment]
> @mike-000 @jahow I may have stumbled upon a more fundamental problem. It appears every layer gets rendered into a separate context.

That will change when #12965 is complete

**Comment by hackermd** [comment]
> > @mike-000 @jahow I may have stumbled upon a more fundamental problem. It appears every layer gets rendered into a separate context.
> 
> That will change when #12965 is complete

Excellent!

I also like @jahow's suggestion to change the blending behavior using event listeners. Together with #12933 we should be all set then, shouldn't we? Or do you have an alternative idea?

**Comment by jahow** [comment]
> Excellent!
> 
> I also like @jahow's suggestion to change the blending behavior using event listeners. Together with #12933 we should be all set then, shouldn't we? Or do you have an alternative idea?

I think this is the best way to go: these events are here to allow you to tweak the webgl context to your liking right before the `draw` call happens.

**Comment by Punzo** [comment]
Hi @mike-000 @hackermd I had some times to check on this and unfortunately I can not change the gl parameters in the pre renderer event. The event is called  here https://github.com/openlayers/openlayers/blob/main/src/ol/renderer/webgl/TileLayer.js#L341 and then the blendFunc values are changed everytime in prepareDraw (https://github.com/openlayers/openlayers/blob/87d37937c58f4781b9b15611b44ce30a44293975/src/ol/webgl/Helper.js#L571-L574) and in the postProcessing passes (https://github.com/openlayers/openlayers/blob/main/src/ol/webgl/PostProcessingPass.js#L268).

In this PR I propose to have some API to pass to the PostProcessingPass the blending varibales.

File: src/ol/webgl/PostProcessingPass.js [hunk 1]
```diff
@@ -265,7 +268,23 @@ class WebGLPostProcessingPass {
...
+      gl.blendEquation(gl.FUNC_ADD);
+    }
+
+    if (
+      glBlendFuncSRC !== null &&
+      glBlendFuncSRC !== null &&
+      glBlendFuncSRC !== undefined &&
+      glBlendFuncSRC !== undefined
+    ) {
+      gl.blendFunc(glBlendFuncSRC, glBlendFuncDST);
```
Comment (MoonE) [inline src/ol/webgl/PostProcessingPass.js:283]: I think this is supposed to check `glBlendFuncDST` instead of only `glBlendFuncSRC`

Comment (Punzo) [reply src/ol/webgl/PostProcessingPass.js:283]: oh yes, definitely!

File: src/ol/renderer/webgl/TileLayer.js [hunk 2]
```diff
@@ -528,10 +525,10 @@ class WebGLTileLayerRenderer extends WebGLLayerRenderer {
     }
 
     this.helper.finalizeDraw(
-      frameState, 
-      tileLayer.styleVariables_.glBlendEquation,
-      tileLayer.styleVariables_.glBlendFuncSRC,
-      tileLayer.styleVariables_.glBlendFuncDST
+      frameState,
```
Comment (MoonE) [inline src/ol/renderer/webgl/TileLayer.js:528]: `Number(undefined)` would now be `NaN`. 
You can typecast the values instead. And please use a lowercase letter for the start of the method name.
```js
/** @type {GLenum|undefined} */ (tileLayer.getStyleVariables().glBlendEquation),
```

File: src/ol/renderer/webgl/TileLayer.js [hunk 3]
```diff
@@ -528,10 +525,10 @@ class WebGLTileLayerRenderer extends WebGLLayerRenderer {
...
+      Number(tileLayer.GetStyleVariables().glBlendEquation),
+      Number(tileLayer.GetStyleVariables().glBlendFuncSRC),
+      Number(tileLayer.GetStyleVariables().glBlendFuncDST)
```
Comment (MoonE) [inline src/ol/renderer/webgl/TileLayer.js]: `Number(undefined)` would now be `NaN`. 
You can typecast the values instead. And please use a lowercase letter for the start of the method name.
```js
/** @type {GLenum|undefined} */ (tileLayer.getStyleVariables().glBlendEquation),
```

Comment (Punzo) [reply src/ol/renderer/webgl/TileLayer.js]: you are right, thanks for the review and catching that.

**Comment by Punzo** [comment]
Hi @MoonE, I have applied your last review, please let me know if any other action is required on my side. Thanks!

**Comment by hackermd** [comment]
Hi @MoonE please let us know if there are any additional changes that you would like us to make to the PR. It would be great if this feature could get included into the next release.

**Comment by tschaub** [comment]
Thanks for the proposed contribution, @Punzo.  I'm hoping we can minimize the changes to the API while still allowing some customization.  In #13119, I've proposed adding a `precompose` (and `postcompose`) event.  This allows access to the rendering context before layers are composed.  As mentioned there, it is only useful for WebGL layers that are adjacent to one another in rendering order.

I'm curious if you are able to try it out and see if it addresses your use case.  You should be able to do something like:
```js
layer.on('precompose', (event) => {
  const gl = event.context;
  gl.blendEquation(gl.FUNC_ADD);
  gl.blendFunc(gl.SRC_COLOR, gl.DST_COLOR);
  // or whatever else
});
```

I haven't yet come up with a good example of this functionality.  If there are any publicly accessible datasets that would make for an interesting example, let me know.

**Comment by Punzo** [comment]
> ```js
> layer.on('precompose', (event) => {
>   const gl = event.context;
>   gl.blendEquation(gl.FUNC_ADD);
>   gl.blendFunc(gl.SRC_COLOR, gl.DST_COLOR);
>   // or whatever else
> });
> ```

Hi @tschaub thank you very much for the reply and PR. I have tested the PR https://github.com/openlayers/openlayers/pull/13119 and it works:

image

regarding the data, I forward the question to @hackermd @fedorov .

**Comment by hackermd** [comment]
@tschaub here is a minimal example for blending two monochromatic images into a pseudo-color image

```javascript
import DataTileSource from 'ol/source/DataTile'
import Map from 'ol/Map'
import TileLayer from 'ol/layer/WebGLTile'
import View from 'ol/View'
import WebGLHelper from 'ol/webgl/Helper'

const createArray = (value) => {
  const w = 256
  const h = 256
  const a = new Uint8Array(h * w);
  for (let i = 0; i < a.length; i++) {
    a[i] = value;
  }
  return a
}

const helper = new WebGLHelper()

const redSource = new DataTileSource({
  loader: function (z, x, y) {
    console.debug(`load red tile (${z}, ${x}, ${y})`)
    if (x % 2) {
      return Promise.resolve(createArray(255));
    } else {
      return Promise.resolve(createArray(0));
    }
  },
  transition: 0,
  bandCount: 1
})

const redLayer = new TileLayer({
  source: redSource,
  style: {
    color: [
      'interpolate',
      ['linear'],
      ['band', 1],
      0,
      [0, 0, 0, 1],
      1,
      [255, 0, 0, 1]
    ]
  }
})
redLayer.helper = helper
redLayer.on('precompose', (event) => {
  const gl = event.context
  gl.blendEquation(gl.FUNC_ADD)
  gl.blendFunc(gl.SRC_COLOR, gl.DST_COLOR)
})

const greenSource = new DataTileSource({
  loader: (z, x, y) => {
    console.debug(`load green tile (${z}, ${x}, ${y})`)
    return Promise.resolve(createArray(255));
  },
  transition: 0,
  bandCount: 1
})

const greenLayer = new TileLayer({
  source: greenSource,
  style: {
    color: [
      'interpolate',
      ['linear'],
      ['band', 1],
      0,
      [0, 0, 0, 1],
      1,
      [0, 255, 0, 1]
    ]
  }
})
greenLayer.helper = helper
greenLayer.on('precompose', (event) => {
  const gl = event.context
  gl.blendEquation(gl.FUNC_ADD)
  gl.blendFunc(gl.SRC_COLOR, gl.DST_COLOR)
})

const map = new Map({
  target: 'map',
  layers: [redLayer, greenLayer],
  view: new View({
    center: [0, 0],
    zoom: 0,
  })
})
```

**Comment by fedorov** [comment]
> I am not 100% sure that the data that I am using for testing are already open-source

@Punzo no, that dataset is not public - not yet.

**Comment by Punzo** [comment]
> > I am not 100% sure that the data that I am using for testing are already open-source
> 
> @Punzo no, that dataset is not public - not yet.

ok thanks for checking!

**Comment by fedorov** [comment]
We do have another dataset that can be shared publicly. If it would help, let me know and I can locate it.

**Comment by tschaub** [comment]
> here is a minimal example for blending two monochromatic images into a pseudo-color image

Thanks, @hackermd.  I extended this to use three (r, g, b) layers and added it as [a rendering test](https://github.com/openlayers/openlayers/blob/6427d1f57a96df4b78a16ab044612eb774600bec/test/rendering/cases/webgl-precompose-event/main.js) in #13119.

expected

I would be tempted to represent something like this as a single, multi-band data tile source instead of multiple layers with custom blending, but I understand that there could be reasons to use multiple layers and maybe other blending scenarios that make this a useful feature.

Unrelated, but note that the `transition` property goes on the layer instead of the source.  Also see [the rendering test case](https://github.com/openlayers/openlayers/blob/6427d1f57a96df4b78a16ab044612eb774600bec/test/rendering/cases/webgl-precompose-event/main.js) for use of the `array` expression for a color (which works with 0-1 floats instead of 0-255 as with the `color` expression, subtle, but can be useful).

**Comment by hackermd** [comment]
Thanks @tschaub

> I would be tempted to represent something like this as a single, multi-band data tile source instead of multiple layers with custom blending, but I understand that there could be reasons to use multiple layers and maybe other blending scenarios that make this a useful feature.

Using one layer with a single source would certainly be simpler. However, we often have > 40 channels and don't want to fetch all of them at once. That's why we handle each channel via a separate layer.


It would be great to add some documentation for the behavior of `WebGLTileLayers`, especially with respect to sharing a canvas between layers using `WebGLHelper` and customizing the rendering pipeline using event handlers. The latter requires an understanding of the different rendering steps, since one needs to know what event is triggered at what point and what effect setting rendering context parameter has at a given point.

**Comment by tschaub** [comment]
> It would be great to add some documentation for the behavior of `WebGLTileLayers`, especially with respect to sharing a canvas between layers using `WebGLHelper` and customizing the rendering pipeline using event handlers. The latter requires an understanding of the different rendering steps, since one needs to know what event is triggered at what point and what effect setting rendering context parameter has at a given point.

I agree we'll need to add docs on all this.  At the same time, we can anticipate continuing to make changes to the implementation.  In my opinion, the WebGL helper (and things like the post-processing pass) should not be part of the stable API.  I'd consider it an oversight (or bug maybe) that the `@api` annotation was added for these in #8951.

**Pull request closed without merging**